crates/kornia-imgproc/src/features/fast.rs

```diff
@@ -1,162 +1,365 @@
+    let mut consecutive_count = 0;
+    let mut curr_response = 0.0;
+
+    if let ControlFlow::Break(_) = (0..15 + n).try_for_each(|l| {
+        if bins[l % 16] == state {
+            consecutive_count += 1;
+            if consecutive_count == n {
+                curr_response = 0.0;
+                circle_intensities.iter().for_each(|m| {
+                    curr_response += (m - curr_pixel).abs();
+                });
+
+                return ControlFlow::Break(());
+            }
         } else {
-            consecutive_brighter = 0;
-            consecutive_darker = 0;
+            consecutive_count = 0;
         }
 
-        if consecutive_brighter >= arc_length || consecutive_darker >= arc_length {
-            return true;
+        ControlFlow::Continue(())
+    }) {
+        return curr_response;
+    }
+
+    0.0
+}
+
+fn get_peak_mask<A: ImageAllocator>(
+    src: &Image<f32, 1, A>,
+    mask: &mut Image<bool, 1, A>,
+    threshold: f32,
+    border_width: usize,
+) {
+    let mask_size = mask.size();
+
+    let src_slice = src.as_slice();
+    let mask_slice = mask.as_slice_mut();
+
+    src_slice
+        .par_iter()
+        .zip(mask_slice)
+        .enumerate()
+        .for_each(|(i, (src, mask))| {
+            let y = i / mask_size.width;
+            let x = i % mask_size.width;
+
+            if x < border_width
+                || x >= mask_size.width.saturating_sub(border_width)
+                || y < border_width
+                || y >= mask_size.height.saturating_sub(border_width)
+            {
+                *mask = false;
+            } else {
+                *mask = *src > threshold;
+            }
+        });
+}
+
+fn get_high_intensity_peaks<A1: ImageAllocator, A2: ImageAllocator>(
+    src: &Image<f32, 1, A1>,
+    mask: &Image<bool, 1, A2>,
+    min_distance: usize,
+    taken: &mut [bool],
+) -> Vec<[usize; 2]> {
+    let src_size = src.size();
+    let width = src_size.width;
+    let height = src_size.height;
+
+    let coords_intensity: Vec<[usize; 2]> = mask
+        .as_slice()
+        .par_iter()
+        .enumerate()
+        .filter_map(|(i, &value)| {
+            if value {
+                let y = i / width;
+                let x = i % width;
+                Some([y, x])
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    let mut result = Vec::new();
+
+    for [y, x] in coords_intensity {
+        let idx = y * width + x;
+
+        // If this location is already suppressed, skip
+        if taken[idx] {
+            continue;
         }
+
+        // Accept this peak
+        result.push([y, x]);
+
+        // Suppress all within min_distance
+        let y0 = y.saturating_sub(min_distance);
+        let y1 = (y + min_distance + 1).min(height);
+        let x0 = x.saturating_sub(min_distance);
+        let x1 = (x + min_distance + 1).min(width);
+
+        (y0..y1)
+            .flat_map(|yy| (x0..x1).map(move |xx| (yy, xx)))
```

Comment:
is there a way to simplify the flat_map/filter/for_each ? looks a bit overcomplicated